### PR TITLE
DB Connections API: add fields method -> column info

### PR DESCRIPTION
**Pull request by elpaso**
Adds to the DB connections API a method to retrieve columns information as `QgsFields`.

I though about implementing a separate struct to hold column information but I think `QgsFields` have all the information we need to implement `alter table` instructions. 
I'm not too worried about the "hidden" columns because they are most of the times not modifiable anyway (geometry, PK etc.).

```python
md = QgsProviderRegistry.instance().providerMetadata('mssql')
conn = md.createConnection(self.uri, {})
fields = conn.fields('qgis_test', 'someData')
self.assertEqual(fields.names(), ['pk', 'cnt', 'name', 'name2', 'num_char'])
```

### CODE REVIEW

**Comment by Gustry** [comment]
That's cool, exactly what I was missing today while I'm migrating plugins to the new database API !

Would it be possible to extend this to a custom query?
```python
metadata = QgsProviderRegistry.instance().providerMetadata('postgres')
connection = metadata.findConnection('foo')
connection.executeSql("SELECT * FROM test;")
```
To get the fields (headers) used in the query ?

Thanks anyway

File: python/core/auto_generated/qgsabstractdatabaseproviderconnection.sip.in [hunk 1]
```diff
@@ -460,6 +462,21 @@ Raises a QgsProviderConnectionException if any errors are encountered.
...
+%Docstring
+Returns the fields of a ``table`` and ``schema``.
+Raises a QgsProviderConnectionException if any errors are encountered.
+
+.. note::
+
+   the default implementation creates a temporary vector layer, providers may
+   choose to override this method for a greater efficiency.
+
+.. versionadded:: 3.18
```
Comment (Gustry) [inline python/core/auto_generated/qgsabstractdatabaseproviderconnection.sip.in:478]:
```suggestion
.. versionadded:: 3.16
```

File: src/core/qgsabstractdatabaseproviderconnection.h [hunk 2]
```diff
@@ -494,6 +496,16 @@ class CORE_EXPORT QgsAbstractDatabaseProviderConnection : public QgsAbstractProv
      */
     virtual QStringList schemas() const SIP_THROW( QgsProviderConnectionException );
 
+    /**
+     * Returns the fields of a \a table and \a schema.
+     * Raises a QgsProviderConnectionException if any errors are encountered.
+     * \note the default implementation creates a temporary vector layer, providers may
+     * choose to override this method for a greater efficiency.
+     * \since QGIS 3.18
```
Comment (Gustry) [inline src/core/qgsabstractdatabaseproviderconnection.h:504]:
```suggestion
     * \since QGIS 3.16
```
Do not wait so long ;-)

Comment (elpaso) [reply src/core/qgsabstractdatabaseproviderconnection.h:504]: yeah, this is actually the first pisce of a more ambitious project but I guess we can merge it soon.

**Comment by elpaso** [comment]
> That's cool, exactly what I was missing today while I'm migrating plugins to the new database API !
> 
> Would it be possible to extend this to a custom query?
> 
> ```python
> metadata = QgsProviderRegistry.instance().providerMetadata('postgres')
> connection = metadata.findConnection('foo')
> connection.executeSql("SELECT * FROM test;")
> ```
> 
> To get the fields (headers) used in the query ?

It would be nice to have but I don't think so: the current scope is meant to retrieve table columns information to show them in the browser tree and to allow operations like dropping a field and (not in scope for now) renaming fields from the browser.

Query layers are out of scope for this and are read-only anyway.

File: src/core/qgsabstractdatabaseproviderconnection.cpp [hunk 3]
```diff
@@ -212,6 +213,18 @@ QList<QgsAbstractDatabaseProviderConnection::TableProperty::GeometryColumnType>
   return mGeometryColumnTypes;
 }
 
+QgsFields QgsAbstractDatabaseProviderConnection::fields( const QString &schema, const QString &tableName ) const
+{
+  QgsVectorLayer vl { tableUri( schema, tableName ), QStringLiteral( "temp_layer" ), mProviderKey };
```
Comment (nyalldawson) [inline src/core/qgsabstractdatabaseproviderconnection.cpp:218]: If you specify LayerOptions here you could avoid some unnecessary work -- e.g. skipCrsValidation should be true, and maybe setting readExtentFromXml to true would also avoid some unneeded calculations

File: src/core/qgsabstractdatabaseproviderconnection.h [hunk 4]
```diff
@@ -506,6 +518,7 @@ class CORE_EXPORT QgsAbstractDatabaseProviderConnection : public QgsAbstractProv
 ///@endcond
 
     Capabilities mCapabilities = nullptr SIP_SKIP;
+    QString mProviderKey;
```
Comment (nyalldawson) [inline src/core/qgsabstractdatabaseproviderconnection.h:521]: Would be useful to expose this via a getter too!

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
> I'm not too worried about the "hidden" columns because they are most of the times not modifiable anyway (geometry, PK etc.).

If this becomes an issue in future we can always add a new LayerOptions flag to say "include all fields"

**Pull request merged**